Adding gpu metrics collection command (#498)
ResearchComputing/Documentation: docs/compute/monitoring-resources.md

# Monitoring Resources

CU Research Computing has two main tools which can help users monitor their HPC resources:
* [Slurm commands](#monitoring-through-slurm-commands): Slurm provides built-in commands that allow users to retrieve usage summaries, job efficiency data, job history, and priority.
* [XDMoD](#xdmod): A web portal for viewing metrics at the system, partition, and user-levels.

## Monitoring Through Slurm Commands

> You can obtain important usage and efficiency metrics directly through Slurm’s built-in commands and answer the following questions:

Reviewer: ```suggestion
You can obtain important usage and efficiency metrics directly through Slurm’s built-in commands and answer questions posed by the below subsections. 
```

* The elapsed times ranged from 0 hours to 1 hour and 48 minutes

### What is my priority?

> 

Reviewer: I think there is an issue in this section. It is not clearly defining that there is a difference between "priority" and "fair share priority". In fact, it often conflates the two. I feel like the information in the "important" admonition "What is "Priority"?" should be at the top before the `levelfs` command and the content organized appropriately.

Finished Job Statistics
(excludes pending, running, and cancelled jobs)
Average CPU Efficiency 19.21%
Average Memory Usage 0.00G
Average Run-time 3.50s
---------------------
```
 Additionally, `seff-array` will display a histogram of the efficiency statistics all of the jobs in the array, separated into 10% increments. For example: 
```text
CPU Efficiency (%)
---------------------
+0.00e+00 - +1.00e+01  [0]
+1.00e+01 - +2.00e+01  [2]  ████████████████████████████████████████
+2.00e+01 - +3.00e+01  [2]  ████████████████████████████████████████
+3.00e+01 - +4.00e+01  [0]
+4.00e+01 - +5.00e+01  [0]
+5.00e+01 - +6.00e+01  [0]
+6.00e+01 - +7.00e+01  [0]
+7.00e+01 - +8.00e+01  [0]
+8.00e+01 - +9.00e+01  [0]
+9.00e+01 - +1.00e+02  [0]
```
The above indicates that all of the jobs displayed less than 30% CPU efficiency, with the jobs evenly split (2/4) between 10%–20% and 20%–30% efficiency. This information will also be displayed for memory and time efficiency.

```{seealso}
If you are not familiar with Job Arrays in SLURM, you can learn more on the ["Scaling Up with job Arrays" page](../running-jobs/job-arrays.md).
```

### How can I check memory and GPU utilization for my jobs?

> The `sacct` command provides detailed accounting information for completed jobs, including CPU, memory, and GPU metrics (e.g., `gpumem`, `gpuutil`). To see a list of available options and usage examples, run `sacct` with no arguments.

Reviewer: ```suggestion
The `sacct` command provides detailed accounting information for completed jobs, including CPU, memory, and GPU metrics (e.g., `gpumem`, `gpuutil`). To see a list of available options and usage examples, run `sacct` with the help argument.
```

     -s, --state:
                   Select jobs based on their current state or the state
                   they were in during the time period given: running (r),
                   completed (cd), failed (f), timeout (to), resizing (rs),
                   deadline (dl) and node_fail (nf).
     -S, --starttime:
                   Select jobs eligible after this time.  Default is
                   00:00:00 of the current day, unless '-s' is set then
                   the default is 'now'.
     Note, valid start/end time formats are...
                   HH:MM[:SS] [AM|PM]
                   MMDD[YY] or MM/DD[/YY] or MM.DD[.YY]
                   MM/DD[/YY]-HH:MM[:SS]
                   YYYY-MM-DD[THH:MM[:SS]]
                   now[{+|-}count[seconds(default)|minutes|hours|days|weeks]]
```
To view the maximum GPU resource usage for a job, the command is:
```
Usage:
sacct -j <jobid> -pno JobID,TRESUsageInMax
```
Arguments and Options:
| Arguments | Description |
| ------------- |------------|
|`-j <jobid>` | Specifies the Slurm job ID you want information about. Here `jobid` is a positional argument. |
|`-n` | Removes the header row from the output. |

> |`-o TRESUsageInMax` | Chooses the output field(s). Here, you’re requesting to display maximum resource usage for a specific jobid. |

Reviewer: ```suggestion
|`-o TJobID,RESUsageInMax` | Chooses the output field(s). Here, you’re requesting to display the JobID and maximum resource usage. |
```

|`gres/gpumem` | Max GPU memory usage. |
|`gres/gpuutil` | Max GPU utilization as a percentage. |
|`mem` | RAM used by the job. |
|`pages` | Number of memory pages used. Typically for advanced monitoring; can often be ignored. |
|`vmem` | Virtual memory used by the job. Includes RAM + swap + memory-mapped files. |

```{note} 
The `sacct` command may display multiple entries for a single job, depending on how it was submitted and its job structure:
- `.batch` step: Appears for jobs submitted via `sbatch` and reflects the actual workload of the batch script. This is typically the most relevant entry for CPU, memory, and GPU usage.
- `.extern` step: A lightweight Slurm setup step that usually shows minimal resource usage.
- Other job steps or tasks: Jobs submitted via `acompile`, `srun`, or as part of job arrays may not have a `.batch` entry, and each step/task may appear separately.

When reviewing resource usage, users should generally focus on the steps that correspond to their actual workload, rather than relying on all entries indiscriminately.
```

Similarly, to view the average GPU resource usage for a job
```
Usage:

> sacct -j <jobid> -pno JobID,TRESUsageInAve --noconvert

Reviewer: ```suggestion
$ sacct -j <jobid> -pno JobID,TRESUsageInAve --noconvert
```

```
Arguments and Options:
| Arguments | Description |
| ------------- |------------|
|`-j <jobid>` | Filters the report to only show information for the specified job ID. |
|`-n` | Removes the header row from the output. |

> |`-o TRESUsageInMax` | Chooses the output field(s). Here, you’re requesting to display maximum resource usage for a specific jobid. |

Reviewer: ```suggestion
|`-o JobID,TRESUsageInAve` | Chooses the output field(s). Here, you’re requesting to display the JobID and average resource usage. |
```

  -j <jobid>            Filters the report to only show information for the specified job ID.
  -n                    Removes the header row from the output.
  -o TRESUsageInAve     Chooses the output field(s). Here, you’re requesting to display average resource usage metrics.
  --noconvert           Prevents conversion of the units (e.g., MB → GB), ensuring raw data remains unchanged.
```
In order to check average GPU metrics for job 18194943, run:
```
$ sacct -j 18194943 -pno JobID,TRESUsageInAve --noconvert
```
This will display the output:
```
--------------------------------------------------------
18194943||
18194943.batch|cpu=00:56:24,energy=0,fs/disk=350130624578,gres/gpumem=40777023488,gres/gpuutil=100,mem=10621726720,pages=3465,vmem=10439041024|
18194943.extern|cpu=00:00:00,energy=0,fs/disk=5273,gres/gpumem=0,gres/gpuutil=0,mem=262144,pages=0,vmem=0|
--------------------------------------------------------
```
This output contains the same fields as in the `TRESUsageInMax` example, but here they represent average usage instead of peak usage.

```{important}
- GPU metrics are currently available only on NVIDIA GPUs. Accessing these metrics requires CUDA 12 or newer.

> - Currently, AMD GPUs, H200s and Blanca H100 GPUs do not support GPU metric collection.

Reviewer: I think the P100s on `blanca-mktg` may not be able to collect this info, but I would launch a job on those and see

If you are not familiar with Job Arrays in SLURM, you can learn more on the ["Scaling Up with job Arrays" page](../running-jobs/job-arrays.md).
```

### How can I check memory and GPU utilization for my jobs?

Type the command name for usage hint:
```
$ sacct -h
```
```
Purpose: This command reports detailed accounting information for completed jobs, 
including CPU, memory, and GPU metrics such as `gpumem` and `gpuutil`.
```
To view the maximum GPU resource usage for a job, the command is:
```
Usage:
sacct -j <jobid> -pno JobID,TRESUsageInMax 

positional arguments:
  jobid                 Job ID to query.

options:
  -j <jobid>            Specifies the Slurm job ID you want information about.
  -n                    Removes the header row from the output.

>   -o TRESUsageInMax     Chooses the output field(s). Here, you’re requesting to display maximum resource usage for a specific jobid.

Reviewer: I do not think this information should be in a code block as that seems to indicate that it is output.

```
In order to check the metrics for job 18194943, run: 
```
$ sacct -j 18194943 -pno JobID,TRESUsageInMax 
```
This will display the output:
```
--------------------------------------------------------
18194943||
18194943.batch|cpu=00:56:24,energy=0,fs/disk=350130624578,gres/gpumem=38888M,gres/gpuutil=100,mem=10372780K,pages=3465,vmem=10194376K|
18194943.extern|cpu=00:00:00,energy=0,fs/disk=5273,gres/gpumem=0,gres/gpuutil=0,mem=256K,pages=0,vmem=0|
--------------------------------------------------------
```
This tells us:
| Variable | Description |
| ------------- |------------|
|`cpu` | Total CPU time consumed (hours:minutes:seconds).|
|`energy` | Energy consumed by the job in arbitrary units (often 0 for systems without energy accounting).|
|`fs/disk` | Amount of disk I/O performed by the job (in bytes). |
|`gres/gpumem` | Peak GPU memory usage. |

> |`gres/gpuutil` | Peak GPU utilization as a percentage. |

Reviewer: ```suggestion
|`gres/gpumem` | Max GPU memory usage. |
|`gres/gpuutil` | Max GPU utilization as a percentage. |
```

|`mem` | RAM used by the job. |
|`pages` | Number of memory pages used. Typically for advanced monitoring; can often be ignored. |
|`vmem` | Virtual memory used by the job. Includes RAM + swap + memory-mapped files. |

```{note} 

> `sacct` shows multiple entries for each job. The primary entry is the one with the `.batch` step, which reflects the actual workload and contains the meaningful resource-usage metrics. The `.extern` entry corresponds to Slurm’s lightweight setup step and will show minimal or negligible usage. Users should reference the `.batch` line when reviewing their job’s resource usage.

Reviewer: It seems like you are making a general statement about `sacct` here that is not exactly true. `batch` will only appear for job submitted via sbatch e.g. it will not show up for `acompile` jobs. Also, this does not account for job arrays or job steps

```

Similarly, to view the average GPU resource usage for a job
```
Usage:

> sacct -j <jobid> -pno JobID,TRESUsageInAve --noconvert

Reviewer: Please split this up so it is its own code block and change the items below as I suggested for the other example above.

options:
  -j <jobid>            Filters the report to only show information for the specified job ID.
  -n                    Removes the header row from the output.
  -o TRESUsageInAve     Chooses the output field(s). Here, you’re requesting to display average resource usage metrics.
  --noconvert           Prevents conversion of the units (e.g., MB → GB), ensuring raw data remains unchanged.
```
In order to check average GPU metrics for job 18194943, run:
```
$ sacct -j 18194943 -pno JobID,TRESUsageInAve --noconvert
```
This will display the output:
```
--------------------------------------------------------
18194943||
18194943.batch|cpu=00:56:24,energy=0,fs/disk=350130624578,gres/gpumem=40777023488,gres/gpuutil=100,mem=10621726720,pages=3465,vmem=10439041024|
18194943.extern|cpu=00:00:00,energy=0,fs/disk=5273,gres/gpumem=0,gres/gpuutil=0,mem=262144,pages=0,vmem=0|
--------------------------------------------------------
```
This output contains the same fields as in the `TRESUsageInMax` example, but here they represent average usage instead of peak usage.

```{important}
- GPU metrics are currently available only on select NVIDIA GPUs and require CUDA 12 or newer.

> - At this time, GPU memory and utilization metrics are not available on the following configurations: AMD GPUs, GH200s and MIG enables GPUs on Alpine, P100s and A40s on Blanca.

Reviewer: ```suggestion
- GPU metrics are available for most NVIDIA GPUs across Alpine and Blanca. At this time, GPU memory and utilization metrics are not available on the following configurations: AMD GPUs, GH200s and MIG enabled GPUs on Alpine, and the P100s and A40s on Blanca.
- GPU metrics can only be collected if your code utilizes CUDA 12 or newer.
```

- Core cluster GPUs (e.g., core-gpu[0-4], viz1, viz2), which power Core Desktop and MATLAB GUI, do not currently support GPU memory or utilization metrics.

> - Users running jobs on unsupported GPUs or older CUDA versions will see zeros or infinite value for GPU memory and utilization fields. Make sure your jobs are running on compatible hardware to obtain meaningful GPU metrics.

Reviewer: ```suggestion
- Users running jobs on unsupported GPUs or older CUDA versions will see zeros or infinite values for GPU memory and utilization fields. Make sure your jobs are running on compatible hardware to obtain meaningful GPU metrics.
```

If you are not familiar with Job Arrays in SLURM, you can learn more on the ["Scaling Up with job Arrays" page](../running-jobs/job-arrays.md).
```

### How can I check memory and GPU utilization for my jobs?

Type the command name for usage hint:
```
$ sacct
```
```
Purpose: This command reports detailed accounting information for completed jobs, 
including CPU, memory, and GPU metrics such as `gpumem` and `gpuutil`.
```
To view the maximum GPU resource usage for a job, the command is:
```
Usage:
sacct -j <jobid> -Pno TRESUsageInMax -p

positional arguments:
  jobid                 Job ID to query.

options:
  -j <jobid>            Specifies the Slurm job ID you want information about.
  -P                    Output in pipe-delimited format (parseable).
  -n                    Removes the header row from the output.
  -o TRESUsageInMax     Chooses the output field(s). Here, you’re requesting to display maximum resource usage.
  -p                    Print fields with trailing delimiters (helpful for scripts).
```
In order to check the metrics for job 18194943, run: 
```

> $ sacct -j 18194943 -Pno TRESUsageInMax -p

Reviewer 2: For a standard job (i.e. one that does use srun) Slurm will provide three lines of data - Job, Batch, and Extern. While the Job's initial line has no data, the empty line will still appear in the output. To help newbies understand which is which, I would recommend including the JobID.

--------------------------------------------------------
```
This tells us:
| Variable | Description |
| ------------- |------------|
|`cpu` | Total CPU time consumed (hours:minutes:seconds).|
|`energy` | Energy consumed by the job in arbitrary units (often 0 for systems without energy accounting).|
|`fs/disk` | Amount of disk I/O performed by the job (in bytes). |
|`gres/gpumem` | Peak GPU memory usage. |
|`gres/gpuutil` | Peak GPU utilization as a percentage. |
|`mem` | RAM used by the job. |
|`pages` | Number of memory pages used. Typically for advanced monitoring; can often be ignored. |
|`vmem` | Virtual memory used by the job. Includes RAM + swap + memory-mapped files. |

```{note} 
`sacct` reports two entries for each job: the first line reflects the actual workload (the primary job step) and contains the meaningful resource-usage metrics, while the second line corresponds to Slurm’s lightweight `extern` step, which performs the job’s shell setup and shows negligible usage. The first line is the one users should reference.
```

Similarly, to view the average GPU resource usage for a job
```
Usage:

> sacct -j <jobid> -Pno TRESUsageInAve -p --noconvert

Reviewer 2: ```suggestion
sacct -j <jobid> -pno JobID,TRESUsageInAve --noconvert
```

positional arguments:
  jobid                 Job ID to query.

options:
  -j <jobid>            Filters the report to only show information for the specified job ID.
  -P                    Output in pipe-delimited format (parseable).
  -n                    Removes the header row from the output.
  -o TRESUsageInAve     Chooses the output field(s). Here, you’re requesting to display average resource usage metrics.
  -p                    Print fields with trailing delimiters (helpful for scripts).
  --noconvert           Prevents conversion of the units (e.g., MB → GB), ensuring raw data remains unchanged.
```
In order to check average GPU metrics for job 18194943, run:
```
$ sacct -j 18194943 -Pno TRESUsageInAve -p --noconvert
```
This will display the output:
```
--------------------------------------------------------
cpu=00:56:24,energy=0,fs/disk=350130624578,gres/gpumem=40777023488,gres/gpuutil=100,mem=10621726720,pages=3465,vmem=10439041024|

> cpu=00:00:00,energy=0,fs/disk=5273,gres/gpumem=0,gres/gpuutil=0,mem=262144,pages=0,vmem=0|

Reviewer 2: ```suggestion
18194943.extern | cpu=00:00:00,energy=0,fs/disk=5273,gres/gpumem=0,gres/gpuutil=0,mem=262144,pages=0,vmem=0|
```
